src/Bicep.Core/Syntax/ForSyntax.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Bicep.Core.Parsing;
+
+namespace Bicep.Core.Syntax
+{
+    public class ForSyntax : SyntaxBase
+    {
+        public ForSyntax(
+            Token openSquare,
+            Token forKeyword,
+            LocalVariableSyntax itemVariable,
+            SyntaxBase inKeyword,
+            SyntaxBase expression,
+            SyntaxBase colon,
+            SyntaxBase body,
+            SyntaxBase closeSquare)
+        {
+            AssertTokenType(openSquare, nameof(openSquare), TokenType.LeftSquare);
+            AssertKeyword(forKeyword, nameof(forKeyword), LanguageConstants.ForKeyword);
+            AssertSyntaxType(inKeyword, nameof(inKeyword), typeof(Token), typeof(SkippedTriviaSyntax));
+            AssertKeyword(inKeyword as Token, nameof(inKeyword), LanguageConstants.InKeyword);
+            AssertSyntaxType(colon, nameof(colon), typeof(Token), typeof(SkippedTriviaSyntax));
+            AssertTokenType(colon as Token, nameof(colon), TokenType.Colon);
+            AssertSyntaxType(closeSquare, nameof(closeSquare), typeof(Token), typeof(SkippedTriviaSyntax));
+            AssertTokenType(closeSquare as Token, nameof(closeSquare), TokenType.RightSquare);
+            
+            this.OpenSquare = openSquare;
+            this.ForKeyword = forKeyword;
+            this.ItemVariable = itemVariable;
+            this.InKeyword = inKeyword;
+            this.Expression = expression;
+            this.Colon = colon;
+            this.Body = body;
+            this.CloseSquare = closeSquare;
+        }
+
+        public Token OpenSquare { get; }
+
+        public Token ForKeyword { get; }
+
+        public LocalVariableSyntax ItemVariable { get; }
```

Comment:
Why not just use `IdentifierSyntax` here?

Reply:
It would make binding to be inconsistent between loops and other declarations. We bind all other declarations to symbols. All the other declarations have `NameSyntax` nodes that aren't bound to anything. It's kind of weird to have some `IdentifierSyntax` nodes be bound but others. It would also require special cases in the nav code in the language server to not ignore `IdentifierSyntax` for some nodes. The nav features rely on `INamedDeclaration` interface currently, which doesn't belong on `IdentifierSyntax` also.

All of this could be refactored, but I also don't see a compelling reason to do so right now. Sure, the tree gets a little deeper, but loops are not going to be all that numerous anyway (given the runtime constraints).